### ISSUE TITLE
torch.norm is numerically unstable at zero for multidim reductions

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#37323]
Note the inconsistent gradient computation:

```
import torch
a = torch.zeros(3, 3, 3, requires_grad = True)
print(torch.autograd.grad(a.norm(dim = 1).sum(), (a,))[0])
print(torch.autograd.grad(a.norm(dim = (1, 2)).sum(), (a,))[0])
print(torch.__version__)
```

```
tensor([[[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]]])
tensor([[[nan, nan, nan],
         [nan, nan, nan],
         [nan, nan, nan]],

        [[nan, nan, nan],
         [nan, nan, nan],
         [nan, nan, nan]],

        [[nan, nan, nan],
         [nan, nan, nan],
         [nan, nan, nan]]])
1.6.0.dev20200417
```

Relevant code: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L558

Relevant discussion: https://github.com/pytorch/pytorch/issues/37272#issuecomment-619405043

cc @ezyang @SsnL @albanD @zou3519 @gqchen @ngimel

**Comment by vadimkantorov** [comment]
There can be other similar bugs where multidim path is different from special-case-handling default path

**Comment by albanD** [comment]
Agree with @vadimkantorov, we should add the same special cases at 0 to these new functions.

**Comment by ngimel** [comment]
torch.norm with multidim reduction is not a new function, this has existed for a long time.

**Comment by albanD** [comment]
Yes, sorry, not new. But added at the same time: https://github.com/pytorch/pytorch/pull/11261
In any case, it should be fixed yes.

**Comment by vadimkantorov** [comment]
It's quite strange that nan happen because norm_backward for IntArray calls into the default norm_backward that removes nans: https://github.com/pytorch/pytorch/blob/master/tools/autograd/templates/Functions.cpp#L135

**Comment by albanD** [comment]
But in the function you shared above: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L552 you can see that the 1dim case calls this function that removed nan, while the other cases do prod/sum/sqrt (and splitting the graph here leads to nan).

**Comment by vadimkantorov** [comment]
Don't they use the custom backward that I linked to?

**Comment by albanD** [comment]
No, `frobenius_norm` is not registered in `derivatives.yaml`. So it is autodiffed.
`norm` is though and is listed there with `norm_backward()`.

You can check with this to see the difference between the two.
```python
import torch
import torchviz

a = torch.zeros(3, 3, 3, requires_grad = True)
one_dim = a.norm(dim = 1).sum()
two_dims = a.norm(dim = (1, 2)).sum()
print(torch.autograd.grad(one_dim, (a,))[0])
print(torch.autograd.grad(two_dims, (a,))[0])
print(torch.__version__)


torchviz.make_dot(one_dim).view()
input()
torchviz.make_dot(two_dims).view()
```

**Comment by vadimkantorov** [comment]
So is https://github.com/pytorch/pytorch/blob/master/tools/autograd/templates/Functions.cpp#L119 dead/uncalled code?

**Comment by albanD** [comment]
The function no, the first one is called when no dim is given, and the one you linked when 1 dim is given:
```
- name: norm.Scalar(Tensor self, Scalar p=2) -> Tensor
  self: norm_backward(grad, self, p, result)

- name: norm.ScalarOpt_dim(Tensor self, Scalar? p, int[1] dim, bool keepdim=False) -> Tensor
  self: norm_backward(grad, self, p, result, dim, keepdim)
```

The `else (dim() > 1)` though.. I guess yes?

**Comment by mrshenli** [comment]
Self-assigned to reserve this for bootcamp

**Comment by mruberry** [comment]
This still occurs with `torch.linalg.norm`.

**Comment by mruberry** [comment]
I was mistaken! @mrshenli pointed out this issue does appear to be fixed with torch.linalg.norm. I must have failed to save my updated version of the script. Here's the script I'm using:

```
import torch
a = torch.zeros(3, 3, 3, requires_grad = True)
print(torch.autograd.grad(torch.linalg.norm(a, dim = (1,)).sum(), (a,))[0])
print(torch.autograd.grad(torch.linalg.norm(a, dim = (1, 2)).sum(), (a,))[0])
print(torch.__version__)
```

And the result:

```
tensor([[[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]]])
tensor([[[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]]])
1.7.0a0+24eea36
```

Since we'll soon be deprecating torch.norm in favor of torch.linalg.norm, I think we can close this issue.

**Comment by vadimkantorov** [comment]
@mruberry Were the code paths for single/multi-dim cases merged into the one?

**Comment by mruberry** [comment]
Check out the PR: https://github.com/pytorch/pytorch/pull/42749.
